README.md
# multi_state_button
[![Pub](https://img.shields.io/badge/null%20safe-%E2%9C%94-brightgreen)](https://pub.dartlang.org/packages/level_map) [![support](https://img.shields.io/badge/github-level__map-brightgreen?style=flat)](https://github.com/Bharathh-Raj/multi_state_button/tree/main/example)
[![support](https://img.shields.io/badge/platform-flutter%7Cflutter%20web%7Cwindows%7Clinux%7Cmac%20os-ff69b4.svg?style=flat)](https://github.com/Bharathh-Raj/level_map)

A package which provides multiple states for a button with endless customizability.

[![Sample](https://i.imgur.com/vxQKlhW.gif)](https://github.com/Bharathh-Raj/multi_state_button)

## Getting Started

### Add dependency

```yaml
dependencies:
  multi_state_button: ^1.1.3
```

### Super simple to use

```dart
import 'package:multi_state_button/multi_state_button.dart';
class MyApp extends StatelessWidget {
    static const String _submit = "Submit";
    static const String _loading = "Loading";
    static const String _success = "Success";
    final MultiStateButtonController multiStateButtonController =  MultiStateButtonController(initialStateName: _submit);
    @override
    Widget build(BuildContext context) {
        return MultiStateButton(
            multiStateButtonController: multiStateButtonController,
            buttonStates:[
                ButtonState(
                  stateName: _submit,
                  child: Text(
                    _submit,
                  ),
                  textStyle: TextStyle(color: Colors.white, fontSize: 20),
                  size: Size(160, 48),
                  color: Colors.blue,
                  onPressed:() => multiStateButtonController.setButtonState = _loading,
                  ),
                  ButtonState(
                  stateName: _loading,
                  alignment: Alignment.center,
                  child: Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: CircularProgressIndicator(
                        strokeWidth: 3,
                        color: Colors.white,
                    ),
                  ),
                  decoration: BoxDecoration(
                    color: Colors.teal,
                    borderRadius: BorderRadius.all(Radius.circular(48)),
                  ),
                  size: Size(48, 48),
                  onPressed: () => multiStateButtonController.setButtonState = _success,
                  ),
                  ButtonState(
                  stateName: _success,
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Text(
                        _success,
                      ),
                      const SizedBox(
                        width: 16,
                      ),
                      Icon(
                        Icons.check,
                        color: Colors.white,
                      )
                    ],
                  ),
                  textStyle: TextStyle(color: Colors.white, fontSize: 22),
                  color: Colors.green,
                  size: Size(200, 48),
                  onPressed: ()=> multiStateButtonController.setButtonState = _submit,
                  ),
            ],
        );
    }
}
```

### Properties to configure

### MultiStateButton properties
```dart
class MultiStateButton extends StatefulWidget {
  /// List of button states.
  /// Note: Default initial State is the first button state.
  /// Can override this behaviour with [initialStateName] constructor parameter of [MultiStateButtonController].
  final List<ButtonState> buttonStates;

  /// Used to change the current state of the button
  final MultiStateButtonController multiStateButtonController;

  /// Transition duration between button state transition.
  /// Defaults to 500 milliseconds
  final Duration transitionDuration;

  /// Transition curve between button state transition.
  /// Defaults to [Curves.easeIn]
  final Curve transitionCurve;
}
```

### SingleStateButton properties
```dart
class SingleStateButton extends StatelessWidget {
  /// State for single state button
  final ButtonState buttonState;
}
```

#### ButtonState Properties
```dart
class ButtonState {
  /// This is the name used to identify the state of the button.
  /// This name is used to change the state using [MultiStateButtonController]
  final String stateName;
  final Widget child;
  final VoidCallback? onPressed;

  /// Size of the button
  final Size? size;

  /// Color of the button.
  /// This should be null if boxDecoration is not null.
  final Color color;

  /// Decorate the button with color, border radius, box shadow, gradient etc...
  final BoxDecoration? decoration;

  /// Used to align the child.
  /// Defaults to [Alignment.center]
  final AlignmentGeometry alignment;

  /// If descendant of the button is a Text child, this textStle is used to enable transition animtation between two states.
  final TextStyle? textStyle;

  /// The transformation matrix to apply before painting the button.
  final Matrix4? transform;

  /// The clip behavior when [decoration] is not null.
  final Clip clipBehavior;

  /// The decoration to paint in front of the child.
  final Decoration? foregroundDecoration;

  /// SubRoutine to run on transition animation ends.
  final VoidCallback? onAnimationEnd;

  /// The alignment of the child, relative to the size of the button, if [transform] is specified.
  /// When [transform] is null, the value of this property is ignored.
  final AlignmentGeometry? transformAlignment;
}
```
## Example Project
[![support](https://img.shields.io/badge/github-level__map-brightgreen?style=flat)](https://github.com/Bharathh-Raj/multi_state_button/tree/main/example)

## Features and bugs
Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/Bharathh-Raj/multi_state_button/issues

CHANGELOG.md
# 1.1.3

- fix: Null textStyle without descendant text widget causes issue(#1)

# 1.1.1

- fix: set default state as first ButtonState for MultiStateButton

# 1.1.0

- feat: added SingleStateButton widget

# 1.0.0

- Initial release

lib/src/button_state.dart
import 'package:flutter/material.dart';
import 'multi_state_button_controller.dart';

class ButtonState {
  /// This is the name used to identify the state of the button.
  /// This name is used to change the state using [MultiStateButtonController]
  final String stateName;
  final Widget child;
  final VoidCallback? onPressed;

  /// Size of the button
  final Size? size;

  /// Color of the button.
  /// This should be null if boxDecoration is not null.
  final Color color;

  /// Decorate the button with color, border radius, box shadow, gradient etc...
  final BoxDecoration? decoration;

  /// Used to align the child.
  /// Defaults to [Alignment.center]
  final AlignmentGeometry alignment;

  /// If descendant of the button is a Text child, this textStle is used to enable transition animtation between two states.
  final TextStyle? textStyle;

  /// The transformation matrix to apply before painting the button.
  final Matrix4? transform;

  /// The clip behavior when [decoration] is not null.
  final Clip clipBehavior;

  /// The decoration to paint in front of the child.
  final Decoration? foregroundDecoration;

  /// SubRoutine to run on transition animation ends.
  final VoidCallback? onAnimationEnd;

  /// The alignment of the child, relative to the size of the button, if [transform] is specified.
  /// When [transform] is null, the value of this property is ignored.
  final AlignmentGeometry? transformAlignment;

  /// Note: For better text transition between states,pass textStyle into [ButtonState.textStyle] property.
  const ButtonState({
    required this.stateName,
    this.child = const SizedBox(),
    this.onPressed,
    this.size,
    this.color = Colors.blue,
    this.decoration,
    this.alignment = Alignment.center,
    this.textStyle,
    this.transform,
    this.clipBehavior = Clip.none,
    this.foregroundDecoration,
    this.onAnimationEnd,
    this.transformAlignment,
  });
}


lib/src/multi_state_button.dart
import 'package:flutter/material.dart';

import '../multi_state_button.dart';
import 'multi_state_button_controller.dart';

class MultiStateButton extends StatefulWidget {
  /// List of button states.
  /// Note: Default initial State is the first button state.
  /// Can override this behaviour with [initialStateName] constructor parameter of [MultiStateButtonController].
  final List<ButtonState> buttonStates;

  /// Used to change the current state of the button
  final MultiStateButtonController multiStateButtonController;

  /// Transition duration between button state transition.
  /// Defaults to 500 milliseconds
  final Duration transitionDuration;

  /// Transition curve between button state transition.
  /// Defaults to [Curves.easeIn]
  final Curve transitionCurve;

  MultiStateButton({
    Key? key,
    required this.buttonStates,
    required this.multiStateButtonController,
    this.transitionDuration = const Duration(milliseconds: 500),
    this.transitionCurve = Curves.easeIn,
  }) : super(key: key);

  @override
  _MultiStateButtonState createState() => _MultiStateButtonState();
}

class _MultiStateButtonState extends State<MultiStateButton>
    with SingleTickerProviderStateMixin {
  late final AnimationController _controller;
  late final Animation<double> _animation;

  @override
  void initState() {
    _controller =
        AnimationController(vsync: this, duration: widget.transitionDuration);
    _animation =
        CurvedAnimation(parent: _controller, curve: widget.transitionCurve);
    _controller.forward(from: 1);
    widget.multiStateButtonController.buttonStateName.addListener(() {
      _controller.forward(from: 0);
    });
    super.initState();
  }

  @override
  void dispose() {
    _controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return ValueListenableBuilder<String>(
      valueListenable: widget.multiStateButtonController.buttonStateName,
      builder: (context, value, child) {
        final ButtonState currentButtonState = value.isEmpty
            ? widget.buttonStates.first
            : widget.buttonStates
                .firstWhere((element) => element.stateName == value);
        return GestureDetector(
          onTap: currentButtonState.onPressed,
          child: AnimatedContainer(
            width: currentButtonState.size?.width,
            height: currentButtonState.size?.height,
            color: currentButtonState.decoration == null
                ? currentButtonState.color
                : null,
            decoration: currentButtonState.decoration,
            alignment: currentButtonState.alignment,
            duration: widget.transitionDuration,
            curve: widget.transitionCurve,
            child: AnimatedDefaultTextStyle(
              duration: widget.transitionDuration,
              curve: widget.transitionCurve,
              // Default theme's text style uses textStyle with inherit is false. Passing it causeing error. Thus overriding it with inherit as true
              style: currentButtonState.textStyle?.copyWith(inherit: true) ??
                  const TextStyle(fontSize: 12),
              child: FadeTransition(
                opacity: _animation,
                child: currentButtonState.child,
              ),
            ),
            clipBehavior: currentButtonState.clipBehavior,
            foregroundDecoration: currentButtonState.foregroundDecoration,
            onEnd: currentButtonState.onAnimationEnd,
            transform: currentButtonState.transform,
            transformAlignment: currentButtonState.transformAlignment,
          ),
        );
      },
    );
  }
}


lib/src/single_state_button.dart
import 'package:flutter/material.dart';
import 'package:multi_state_button/multi_state_button.dart';

class SingleStateButton extends StatelessWidget {
  /// State for single state button
  final ButtonState buttonState;
  const SingleStateButton({Key? key, required this.buttonState})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
        onTap: buttonState.onPressed,
        child: Container(
          width: buttonState.size?.width,
          height: buttonState.size?.height,
          color: buttonState.decoration == null ? buttonState.color : null,
          decoration: buttonState.decoration,
          alignment: buttonState.alignment,
          clipBehavior: buttonState.clipBehavior,
          foregroundDecoration: buttonState.foregroundDecoration,
          transform: buttonState.transform,
          transformAlignment: buttonState.transformAlignment,
          child: buttonState.textStyle != null
              ? DefaultTextStyle(
                  style: buttonState.textStyle!, child: buttonState.child)
              : buttonState.child,
        ));
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:multi_state_button/multi_state_button.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  static const String _submit = "Submit";
  static const String _loading = "Loading";
  static const String _tryAgain = "Try Again";
  static const String _error = "Error";
  static const String _block = "Block";
  static const String _success = "Success";
  static const String _tick = "Tick";
  final MultiStateButtonController multiStateButtonController = MultiStateButtonController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Multi State Button Example",
      home: Scaffold(
        appBar: AppBar(title: const Text("Multi State Button Example")),
        body: Center(
          child: MultiStateButton(
            multiStateButtonController: multiStateButtonController,
            buttonStates: [
              ButtonState(
                  stateName: _submit,
                  child: Text(
                    _submit,
                  ),
                  textStyle: TextStyle(color: Colors.white, fontSize: 20),
                  size: Size(160, 48),
                  decoration: BoxDecoration(
                    color: Colors.blue,
                    borderRadius: BorderRadius.all(Radius.circular(24)),
                    boxShadow: <BoxShadow>[
                      BoxShadow(
                        color: Color(0x66666666),
                        blurRadius: 4.0,
                        spreadRadius: 3.0,
                        offset: Offset(0, 2.0),
                      )
                    ],
                  ),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _loading;
                  }),
              ButtonState(
                  stateName: _loading,
                  alignment: Alignment.center,
                  child: Padding(
                    padding: const EdgeInsets.all(8.0),
                    child: SizedBox(
                      height: 24,
                      width: 24,
                      child: CircularProgressIndicator(
                        strokeWidth: 3,
                        color: Colors.white,
                      ),
                    ),
                  ),
                  decoration: BoxDecoration(
                    color: Colors.teal,
                    borderRadius: BorderRadius.all(Radius.circular(48)),
                  ),
                  size: Size(48, 48),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _tryAgain;
                  }),
              ButtonState(
                  stateName: _tryAgain,
                  decoration: BoxDecoration(
                    color: Colors.orange,
                    borderRadius: BorderRadius.all(Radius.circular(24)),
                    boxShadow: const <BoxShadow>[
                      BoxShadow(
                        color: Color(0x66666666),
                        blurRadius: 8.0,
                        spreadRadius: 3.0,
                        offset: Offset(0, 6.0),
                      )
                    ],
                  ),
                  child: Text(
                    _tryAgain,
                  ),
                  size: Size(160, 48),
                  textStyle: TextStyle(color: Colors.white, fontSize: 20, fontWeight: FontWeight.w500),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _error;
                  }),
              ButtonState(
                  stateName: _error,
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Text(
                        _error,
                      ),
                      const SizedBox(
                        width: 16,
                      ),
                      Icon(
                        Icons.error_outline,
                        color: Colors.white,
                      )
                    ],
                  ),
                  textStyle: TextStyle(color: Colors.white, fontSize: 22, fontWeight: FontWeight.bold),
                  decoration: BoxDecoration(
                    color: Colors.red,
                    borderRadius: BorderRadius.all(Radius.circular(12)),
                    boxShadow: <BoxShadow>[
                      BoxShadow(
                        color: Colors.red.withOpacity(0.5),
                        blurRadius: 8.0,
                        spreadRadius: 3.0,
                      )
                    ],
                  ),
                  size: Size(180, 48),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _success;
                  }),
              ButtonState(
                  stateName: _success,
                  child: Row(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Text(
                        _success,
                      ),
                      const SizedBox(
                        width: 16,
                      ),
                      Icon(
                        Icons.check,
                        color: Colors.white,
                      )
                    ],
                  ),
                  textStyle: TextStyle(color: Colors.white, fontSize: 22),
                  decoration: BoxDecoration(
                    color: Colors.green,
                    borderRadius: BorderRadius.all(Radius.circular(48)),
                    boxShadow: const <BoxShadow>[
                      BoxShadow(
                        color: Color(0x66666666),
                        blurRadius: 6.0,
                        spreadRadius: 3.0,
                        offset: Offset(0, -3.0),
                      )
                    ],
                  ),
                  size: Size(200, 48),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _block;
                  }),
              ButtonState(
                  stateName: _block,
                  child: Icon(
                    Icons.block,
                    color: Colors.white,
                  ),
                  decoration: BoxDecoration(
                    color: Colors.blueGrey,
                    borderRadius: BorderRadius.all(Radius.circular(12)),
                  ),
                  size: Size(48, 48),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _tick;
                  }),
              ButtonState(
                  stateName: _tick,
                  child: Icon(
                    Icons.done_rounded,
                    color: Colors.white,
                  ),
                  decoration: BoxDecoration(
                    color: Colors.cyan,
                    borderRadius: BorderRadius.all(Radius.circular(24)),
                    boxShadow: <BoxShadow>[
                      BoxShadow(
                        color: Colors.cyan.withOpacity(0.5),
                        blurRadius: 6.0,
                        spreadRadius: 3.0,
                      )
                    ],
                  ),
                  size: Size(48, 48),
                  onPressed: () {
                    multiStateButtonController.setButtonState = _submit;
                  }),
            ],
          ),
        ),
      ),
    );
  }
}
